src/app/qgisapp.cpp

```diff
@@ -10837,6 +10838,41 @@ void QgisApp::projectCrsChanged()
   }
 }
 
+void QgisApp::projectTemporalRangeChanged()
+{
+  QMap<QString, QgsMapLayer *> layers = QgsProject::instance()->mapLayers();
+  QgsMapLayer *currentLayer = nullptr;
+
+  for ( QMap<QString, QgsMapLayer *>::const_iterator it = layers.constBegin(); it != layers.constEnd(); ++it )
+  {
+    currentLayer = it.value();
+
+    if ( currentLayer->dataProvider() )
+    {
+      QgsDataSourceUri uri;
+      uri.setEncodedUri( currentLayer->dataProvider()->dataSourceUri() );
```

Comment:
Please do not use setEncodedUri! Instead:
QgsDataSourceUri uri { currentLayer->dataProvider()->uri() };

... then:

currentLayer->setDataSource( uri.uri(), currentLayer->name(), currentLayer->providerType(), QgsDataProvider::ProviderOptions() );

Reply:
Hi @elpaso, thanks, the issue is if I don't use setEncodedUri, the URI is not initialized properly. I have tried the way you have suggested many times, but still it doesn't work for WMS URI params parsing. I'll like to also use the QgsDataSourceUri constructor for initialization as long as it parse all the params from WMS.

Reply:
Encoded URI won't work in all cases. So if there is a bug in the `QgsDataSourceUri` constructor you should try to fix it with high priority.

If you can write a test case to demonstrate your issue with WMS URI params parsing, that would be helpful.

Reply:
Ok @elpaso thanks, also I have realized that WMS provider uses Encoded URI everywhere it initialize the provider URI, this means I have to update all those other uses to use the QgsDataSourceURI constructor right?
As of now the updates on WMS-T layers temporal properties is not possible, eg. in this issue https://github.com/qgis/QGIS/issues/3593

Reply:
> Ok @elpaso thanks, also I have realized that WMS provider uses Encoded URI everywhere it initialize the provider URI, this means I have to update all those other uses to use the QgsDataSourceURI constructor right?

If you mean internally in the WMS provider code, I don't see why. 

The thing here is that you must write your code (outside a specific provider) in a way that is not bound to a particular provider (beware that it's not your fault, we all know `QgsDataSourceUri` is a mess).

So, either use the way I recommended you a few lines above (never use encoded uri!) or fix `QgsDataSourceUri` encode/decode to support cases where the provider source string is **not** in encoded uri form (such as postgres and almost all providers except for OGC services).

> As of now the updates on WMS-T layers temporal properties is not possible, eg. in this issue #3593

I could not reproduce that one, sorry.

Btw, I 'm still waiting for a python code snippet to reproduce your claim that "... the issue is if I don't use setEncodedUri, the URI is not initialized properly. ".

Reply:
About the python code snippet, I'm attaching one below, using this WMS Layer https://geo.weather.gc.ca/geomet?layer=GDPS.ETA_TT

`layer = iface.activeLayer()`
`encoded_uri = QgsDataSourceUri()`
`encoded_uri.setEncodedUri(layer.dataProvider().dataSourceUri())`

`# This below returns the right value 'EPSG:4326'`
`print(encoded_uri.param("crs") )`

`normal_uri = QgsDataSourceUri(layer.dataProvider().uri())`

`# This below returns a string of params with their values instead of a crs string,`
`# 'EPSG:4326&dpiMode=7&enableTime=yes&format=image/png&layers=GDPS.ETA_TT&...'`
`print(normal_uri.param("crs"))`

Reply:
> If you mean internally in the WMS provider code, I don't see why.

So as to have only one way of initializing the QgsDataSourceURI, and not mixing two different ways (using constructor externally while using setEncodedUri function internally).

Reply:
I see, QgsDataSourceUri is totally broken we should definitely bypass it and use encode/decodeUri from provider metadata.

I can take care of the implementation for PG raster, can you make it for WMS?

Here is a quick and dirty sample implementation (need to be tested!)

```c++
QVariantMap QgsWmsProviderMetadata::decodeUri(const QString& uri)
{
  const QUrlQuery query { uri };
  const auto constItems { query.queryItems() };
  QVariantMap decoded;
  for (const auto &item: constItems )
  {
    decoded[ item.first ] = item.second;
  }
  return decoded;
}

QString QgsWmsProviderMetadata::encodeUri(const QVariantMap& parts)
{
  QUrlQuery query;
  QList<QPair<QString, QString> > items;
  for (auto it = parts.constBegin(); it != parts.constEnd(); ++it )
  {
    items.push_back({it.key(), it.value().toString() });
  }
  query.setQueryItems( items );
  return query.toString();
}
```

This is how it is supposed to work in a provider agnostic way:

```python
>> md = QgsProviderRegistry.instance().providerMetadata('postgresraster')
# The following uri comes directly from ayer.dataProvider().dataSourceUri().
>> uri = 'dbname=\'qgis_tests\' host=localhost port=5432 user=\'ale\' password=\'ale\' sslmode=disable estimatedmetadata=true srid=0 temporalDefaultTime=\'2020-04-23T13:51:00\' temporalFieldIndex=\'2\' table="idro"."hn_giornaliera" (rast)'
>> decoded = md.decodeUri(uri)
>> print(decoded)
decoded = {'dbname': 'qgis_tests',
 'estimatedmetadata': True,
 'geometrycolumn': 'rast',
 'host': 'localhost',
 'password': 'ale',
 'port': '5432',
 'schema': 'idro',
 'selectatid': False,
 'srid': '0',
 'sslmode': 1,
 'table': 'hn_giornaliera',
 'username': 'ale'}
```
Now, suppose you want to set `enableTime`:

```python
>> decoded['enableTime'] = True  # be careful here: better use a bool than yes/no!
>> print(md.encodeUri(decoded))
'dbname=\'qgis_tests\' user=\'ale\' password=\'ale\' srid=0 enableTime=\'true\' estimatedmetadata=\'true\' host=\'localhost\' key=\'\' port=\'5432\' selectatid=\'false\' service=\'\' sslmode=\'disable\' temporalDefaultTime=\'2020-04-23T13:51:00\' temporalFieldIndex=\'2\' table="idro"."hn_giornaliera" (rast)'
```

The idea is to delegate to the provider the way it manages the uri.


So, for WMS, it would work the same:

```python
In [1]: layer = iface.activeLayer()

In [2]: md = QgsProviderRegistry.instance().providerMetadata('wms')

In [3]: md.decodeUri(layer.dataProvider().dataSourceUri())
Out[3]: 
{'contextualWMSLegend': '0',
 'crs': 'EPSG:4326',
 'dpiMode': '7',
 'enableTime': 'yes',
 'featureCount': '10',
 'format': 'image/png',
 'layers': 'GDPS.ETA_TT',
 'referenceTimeDimensionExtent': '2020-04-21T12:00:00Z/2020-04-23T00:00:00Z/PT12H',
 'styles': NULL,
 'temporalSource': 'provider',
 'timeDimensionExtent': '2020-04-23T00:00:00Z/2020-05-03T00:00:00Z/PT3H',
 'type': 'wmst',
 'url': 'https://geo.weather.gc.ca/geomet?layer=GDPS.ETA_TT'}
In [4]: decoded = md.decodeUri(layer.dataProvider().dataSourceUri())
In [5]: decoded['enableTime'] = False

In [7]: md.encodeUri(decoded)
Out[7]: 'contextualWMSLegend=0&crs=EPSG:4326&dpiMode=7&enableTime=false&featureCount=10&format=image/png&layers=GDPS.ETA_TT&referenceTimeDimensionExtent=2020-04-21T12:00:00Z/2020-04-23T00:00:00Z/PT12H&styles&temporalSource=provider&timeDimensionExtent=2020-04-23T00:00:00Z/2020-05-03T00:00:00Z/PT3H&type=wmst&url=https://geo.weather.gc.ca/geomet?layer%3DGDPS.ETA_TT'

```

Reply:
@Samweli  see: https://github.com/qgis/QGIS/pull/35952

Reply:
Cool!, How about saving the provider URI after adding/updating its params? Won't it affect the past structure of the URI when saving new changes using this approach?

Reply:
> @Samweli see: #35952

Ok

Reply:
> Cool!, How about saving the provider URI after adding/updating its params? Won't it affect the past structure of the URI when saving new changes using this approach?

Not sure if I understand what you mean, but if you want to call `setDataSource()` on the layer, it takes a string representation of the uri and it should work without any conversion.

In any event, the encode/decode is controlled by the provider through provider's metadata, it's the provider responsibility to decode/encode correctly, a round trip test it's probably the best check.

Reply:
Ok, so with these updates how will the following code change to..

`QgsDataSourceUri uri { currentLayer->dataProvider()->uri() };`

... then:

`currentLayer->setDataSource( uri.uri(), currentLayer->name(), currentLayer->providerType(), QgsDataProvider::ProviderOptions() );`

Reply:
Thanks very much for all this great input @elpaso !